pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1564,3 +1587,85 @@ func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd
 	}
 	return true, nil
 }
+
+func calculateNextProvisionTime(provision *hivev1.ClusterProvision, cdLog log.FieldLogger) time.Time {
+	cond := controllerutils.FindClusterProvisionCondition(provision.Status.Conditions, hivev1.ClusterProvisionFailedCondition)
+	if cond == nil || cond.Status != corev1.ConditionTrue {
+		cdLog.Warnf("failed provision does not have a %s condition", hivev1.ClusterProvisionFailedCondition)
+		return time.Now()
+	}
+	failTime := cond.LastTransitionTime
+
+	// (2^currentRetries) * 60 seconds up to a max of 24 hours.
+	const sleepCap = 24 * time.Hour
+	const retryCap = 11 // log_2_(24*60)
+
+	retries := provision.Spec.Attempt
+	if retries >= retryCap {
+		return failTime.Add(sleepCap)
+	}
+	return failTime.Add((1 << uint(retries)) * time.Minute)
+}
+
+func (r *ReconcileClusterDeployment) existingProvisions(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) ([]*hivev1.ClusterProvision, error) {
+	provisionList := &hivev1.ClusterProvisionList{}
+	if err := r.List(
+		context.TODO(),
+		provisionList,
+		client.InNamespace(cd.Namespace),
+		client.MatchingLabels(map[string]string{clusterprovision.ClusterDeploymentLabelKey: cd.Name}),
+	); err != nil {
+		cdLog.WithError(err).Warn("could not list provisions for clusterdeployment")
+		return nil, err
+	}
+	provisions := make([]*hivev1.ClusterProvision, len(provisionList.Items))
+	for i := range provisionList.Items {
+		provisions[i] = &provisionList.Items[i]
+	}
+	return provisions, nil
+}
+
+func (r *ReconcileClusterDeployment) getFirstProvision(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) *hivev1.ClusterProvision {
+	provisions, err := r.existingProvisions(cd, cdLog)
+	if err != nil {
+		return nil
+	}
+	for _, provision := range provisions {
+		if provision.Spec.Attempt == 0 {
+			return provision
+		}
+	}
+	cdLog.Warn("could not find the first provision for clusterdeployment")
+	return nil
+}
+
+func (r *ReconcileClusterDeployment) adoptOrphanedProvision(cd *hivev1.ClusterDeployment, provision *hivev1.ClusterProvision, cdLog log.FieldLogger) (reconcile.Result, error) {
+	pLog := cdLog.WithField("provision", provision.Name)
+	pLog.Error("found orphaned provision with an active job")
+	cd.Status.Provision = &corev1.LocalObjectReference{Name: provision.Name}
+	if err := r.Status().Update(context.TODO(), cd); err != nil {
+		pLog.WithError(err).Error("could not adopt orphaned provision")
+		return reconcile.Result{}, err
+	}
+	pLog.Info("adopted orphaned provision")
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileClusterDeployment) deleteStaleProvisions(provs []*hivev1.ClusterProvision, cdLog log.FieldLogger) {
+	// Cap the number of existing provisions. Always keep the earliest provision as
+	// it is used to determine the total time that it took to install. Take off
+	// one extra to make room for the new provision being started.
```

Comment:
Ok this is awesome, I can't remember mentioning it but I thought first + a few most recent would be ideal but too fussy to request.  +100